File: go/011.go
package main

/**
给定一个长度为 n 的整数数组 height 。
有 n 条垂线，第 i 条线的两个端点是 (i, 0) 和 (i, height[i]) 。
找出其中的两条线，使得它们与 x 轴共同构成的容器可以容纳最多的水。
返回容器可以储存的最大水量。
思路：双指针 + 贪心，从两边向中间靠拢，取最大值即可。移动小的指针（短板效应）。
*/
func maxArea(height []int) int {
	res := 0
	for i, j := 0, len(height)-1; i < j; {
		h, l := 0, j-i
		if height[i] < height[j] {
			h = height[i]
			i++
		} else {
			h = height[j]
			j--
		}
		area := l * h
		res = max(area, res)
	}
	return res
}

File: c++/011.cpp
class Solution {
/*
若向内 移动短板 ，水槽的短板 min(h[i], h[j]) 可能变大，因此下个水槽的面积 可能增大 。
若向内 移动长板 ，水槽的短板 min(h[i], h[j])不变或变小，因此下个水槽的面积 一定变小 。
因此，初始化双指针分列水槽左右两端，循环每轮将短板向内移动一格，并更新面积最大值，直到两指针相遇时跳出；即可获得最大面积。
*/
public:
    int maxArea(vector<int>& height) {
        int res=0;
        for(int i=0, j=height.size()-1; i<j;){
            res = max(res,(j-i)*min(height[i],height[j]));
            height[i]>height[j]?j--:i++;
        }
        return res;
    }
};